Replace all LVGL driver I2C code with I2C Manager

ropg opened the pull request:
This PR integrates my I2C Manager code into lvgl_esp32_drivers. It replaces various bits and pieces of I2C code (which was a bit ... ehm ... organically grown) with one code-base for all I2C comms. This makes everything more modular and more readable. I2C manager is thread-safe, which is important when others, from other threads, want frequent access to I2C while LVGL needs to talk to a touchpad or display many times a second, because I2C using ESP-IDF is not thread-safe. 

#### Single submenu

There is now one submenu for the drivers that has a TFT, a Touch and optionally an I2C submenu:

```
menu "LVGL ESP Drivers"

    rsource "lvgl_tft/Kconfig"

    rsource "lvgl_touch/Kconfig"

    rsource "i2c_manager/Kconfig"

endmenu
```

&nbsp;

Below is the README file that is in the `i2c_manager` subdirectory. 

## Information for users

### I2C Manager support

`lvgl_esp32_drivers` comes with built-in I2C support by integrating I2C Manager, which is used in case your touch interface or screen uses the I2C bus. The native I2C support offered by ESP-IDF is not thread-safe. Maybe you use LVGL with a touch sensor that has an i2c port, and maybe your device also has another i2c device that needs to be read frequently, such as a 3D-accelerometer. If you read that from another task than the lvgl uses to read the touch data, you need some kind of mechanism to keep these communications from interfering. 

If you have other components that can use I2C Manager (or Mika Tuupola's I2C HAL abstraction that I2C Manager is compatible with) then put I2C Manager in your components directory by cloning the repository from below and in your main program do:

```c
#include "i2c_manager.h"
#include "lvgl_helpers.h"

[...]

lvgl_locking(i2c_manager_locking());
lv_init();
lvgl_driver_init();
```

The `lvgl_locking` part will cause the LVGL I2C driver to play nice with anything else that uses the I2C port(s) through I2C Manager.

See the [I2C Manager GitHub repository](https://github.com/ropg/i2c_manager) for much more information. 

&nbsp;

## Information for driver developers

I2C support in the LVGL ESP drivers is provided exclusively by the files in this directory. Code from all over the project that was talking to the I2C hardware directly has been replaced by code that communicates through the functions provided in `lvgl_i2c.h`. I2C is handled by the I2C Manager that was built into lvlg_esp32_drivers, but the code would be the same if it was routed through I2C Manager as a separate component. If you are providing a driver, you need not worry about any of this.

### Using I2C in a driver, a multi-step guide

#### Step 1

The Kconfig entries for your driver only need to specify that you will be using I2C. This is done by `select LV_I2C_DISPLAY` or `select LV_I2C_TOUCH`.

#### Step 2

To use the I2C port in your code you would do something like:

```c
#include "i2c_manager/i2c_manager.h"

uint8_t data[2];
lvgl_i2c_read(CONFIG_LV_I2C_TOUCH_PORT, 0x23, 0x42, &data, 2);
```

This causes a touch driver to read two bytes at register `0x42` from the IC at address `0x23`. Replace `CONFIG_LV_I2C_TOUCH_PORT` by `CONFIG_LV_I2C_DISPLAY_PORT` when this is a display instead of a touch driver. `lvgl_i2c_write` works much the same way, except writing the bytes from the buffer instead of reading them.

> The example above ignores it but these functions return `esp_err_t` so you can check if the i2c communication worked.

#### Step 3

There is no step 3, you are already done.

### Behind the scenes

If anything in `lvgl_esp32_drivers` uses I2C, the config system will pop up an extra menu under the lvgl drivers menu. This will allow you to select an I2C port for screen and one for the touch driver, if applicable. An extra menu allows you to set the GPIO pins and bus speed of any port you have selected for use. It's perfectly fine for a display and a touch driver to use the same I2C port or different ones. 


## More information

If you need more documentation, please refer to the [I2C Manager GitHub repository](https://github.com/ropg/i2c_manager) for more detailed information on how I2C manager works.

Closes https://github.com/lvgl/lv_port_esp32/issues/282

C47D commented on lvgl_tft/st7735s.c:
```diff
@@ -163,12 +164,15 @@ void st7735s_flush(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * col
-}
+    #ifdef CONFIG_LV_M5STICKC_HANDLE_AXP192
+    	axp192_sleep_in();
+    #endif}
```
`}` should be in a new line.

ropg replied:
whoops, good catch... Fixed.

C47D commented:
> (which was a bit ... ehm ... organically grown)

Totally agree 😄 , life happened and I didn't had time to work on clean it up.

I requested the review of @tore-espressif as he is more familiar with ESP32 and I don't have time until weekend to review this and @kisvegabor and @embeddedt because they know the future work we're planning with this repo.

Thanks for the efford you've placed into this  @ropg 😸

tore-espressif commented:
First off all, thank you very much for you thorough work. Thread-safety is something that we definitely need for I2C.

Before I get to the real code-review (beginning of next week?) I'd like to ask you some higher-level questions:

1. what are the long term plans with https://github.com/ropg/i2c_manager ? could we use it as a submodule?
2. I can see that you have significant contributions in LVGL, would you mind having a look on my refactoring proposal? https://github.com/lvgl/lvgl_esp32_drivers/issues/72. Mostly the section about Kconfig, which I'd like to wipe out from this repo :)

ropg commented:
> 1. what are the long term plans with https://github.com/ropg/i2c_manager ? could we use it as a submodule?

Presently I2C Manager has a few different ways of using it, the README has lots of details. I built it such that it can be a component by itself, or it can be built into a component. The problem with ESP-IDF that I found is that you CANNOT do conditional dependencies, as all dependencies are resolved before any of the Kconfig stuff is parsed. So depending on it as a component means everyone has to install it, even if no I2C is used.

That leaves incorporating it in our component. I built it such that one only has to incorporate the inner directory (also called `i2c_manager`), not the repo root, and I just copied the files. (It's just 4 files: .c, .h, Kconfig and README). It's important that these files do net end up in a dir that is in INCLUDE_DIRS, as it is only to be included from this component. I would prefer it not to be a submodule as I'd have to change how it works a bit then. (Presently the .h file has `#define I2C_OEM lvgl` in it which tells I2C Manager what some functions are called.) I'll make sure that if anything were to change I update the version here also.

*(If people want to get thread-safety between something they are doing and LVGL they need to install and use the I2C Manager component and tell the version incorporated in LVGL to use its locking semaphore pointers, which is in the README.)*

&nbsp;

> 2. I can see that you have contributions in LVGL, would you mind having a look on my refactoring proposal? #72. Mostly the section about Kconfig, which I'd like to wipe out from this repo :)

Will do and will respond there.

C47D commented:
#69 just got merged, I guess the changes it did to the conflicted Kconfig files are making this PR not merging cleanly, will try to take a look at those changes, it also contained (IIRC) I2C related code.

-Carlos

tore-espressif commented on i2c_manager/Kconfig:
```diff
@@ -0,0 +1,112 @@
+		if I2C_MANAGER_0_ENABLED
+			config I2C_MANAGER_0_SDA
+				int "SDA (GPIO pin)"
+				range 0 39 if IDF_TARGET_ESP32
+				range 0 43 if IDF_TARGET_ESP32S2
```
There is comment below about ESP pins that applies here too.

tore-espressif reviewed (commented):
* There are few nitpicks but nothing too important 
* Although I didn't have a chance to try it on HW
* Well structured code that cleans our mess in I2C

In light of comment https://github.com/lvgl/lvgl_esp32_drivers/issues/72#issuecomment-865256557 I'm 100% OK with merging this into our current master, the so to be version 1.0.0.
But unless I'm brutally outvoted, I wouldn't recommend using this in version 2.0.0 and higher.

ropg commented on lvgl_tft/ssd1306.c:
```diff
@@ -240,17 +240,19 @@ static uint8_t send_pixels(lv_disp_drv_t *disp_drv, void *color_buffer, size_t b
-    i2c_cmd_link_delete(cmd);
+    err = lvgl_i2c_write(OLED_I2C_PORT, OLED_I2C_ADDRESS, OLED_CONTROL_BYTE_DATA_STREAM, color_buffer, buffer_len);
+
+//     i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+//     i2c_master_start(cmd);
+//     i2c_master_write_byte(cmd, (OLED_I2C_ADDRESS << 1) | I2C_MASTER_WRITE, true);
+//
+//     i2c_master_write_byte(cmd, OLED_CONTROL_BYTE_DATA_STREAM, true);
+//     i2c_master_write(cmd, (uint8_t *) color_buffer, buffer_len, true);
+//     i2c_master_stop(cmd);
+//
+//     /* Send queued commands */
+//     err = i2c_master_cmd_begin(DISP_I2C_PORT, cmd, 10 / portTICK_PERIOD_MS);
+//     i2c_cmd_link_delete(cmd);
```
I don't actually have an ssd1306. I2C code is fairly straigtforward, nevertheless, I figured I'd leave the original code that interacts with the I2C there until the new code is tested.

C47D replied:
I got one, I can test it later today.

C47D replied:
Fixed in master in https://github.com/lvgl/lvgl_esp32_drivers/commit/b70d2dc151db61ec44cb5b7a2a043917a1d181a9

tore-espressif replied:
I'll try it on HW in a week. If it works well we can remove the commented-out old implementation.

tore-espressif replied:
Works well with commit `9727239e307eb0ee00e54cf7591b02946e41f3fb` and this patch https://github.com/lvgl/lvgl_esp32_drivers/pull/70#discussion_r659406183

Sorry for the late response, I had to work on something else for few days.

ropg commented:
> In light of comment #72 (comment) I'm 100% OK with merging this into our current master, the so to be version 1.0.0.
But unless I'm brutally outvoted, I wouldn't recommend using this in version 2.0.0 and higher.

Naturally I like I2C Manager because I built it. But I don't get too attached to any of my projects, and if something else is better for LVGL, naturally it should be used instead. 

That said:

* Can you expand on why you wouldn't recommend it? It's sort of left hanging, and I'd like to convince you otherwise.

* How would it need to change to be suitable?

or

* What do you recommend in its place?

ropg commented:
@dastarling Your gt911 driver uses I2C. This Pull Request takes out all driver I2C code and replaces with shared I2C code. I'd be happy to refactor your code unless you'd like to. I'd also be happy to add in the multi-byte reads that you have in a TODO as well.

C47D commented:
Got the following error (after updating master with the last commit):

```commandline
../components/lvgl_esp32_drivers/lvgl_tft/ssd1306.c: In function 'send_data':
../components/lvgl_esp32_drivers/lvgl_tft/ssd1306.c:218:59: error: invalid type argument of unary '*' (have 'int')
     err = lvgl_i2c_write(OLED_I2C_PORT, OLED_I2C_ADDRESS, *data[0], data + 1, bytes_len - 1 );
                                                           ^~~~~~~~
[1076/1084] Building C object esp-idf/lvgl_esp32_drivers/CMakeFiles/__idf_lvgl_esp32_drivers.dir/lvgl_helpers.c.ob
ninja: build stopped: subcommand failed.
ninja failed with exit code 1
```

Fixed it with:

```diff
diff --git a/lvgl_tft/ssd1306.c b/lvgl_tft/ssd1306.c
index c806feb..21231c4 100644
--- a/lvgl_tft/ssd1306.c
+++ b/lvgl_tft/ssd1306.c
@@ -215,7 +215,7 @@ static uint8_t send_data(lv_disp_drv_t *disp_drv, void *bytes, size_t bytes_len)

     uint8_t *data = (uint8_t *) bytes;

-    err = lvgl_i2c_write(OLED_I2C_PORT, OLED_I2C_ADDRESS, *data[0], data + 1, bytes_len - 1 );
+    err = lvgl_i2c_write(OLED_I2C_PORT, OLED_I2C_ADDRESS, data[0], data + 1, bytes_len - 1 );

 //     i2c_cmd_handle_t cmd = i2c_cmd_link_create();
 //
```

@ropg I don't have access to the hardware right now, is that the right fix?

ropg commented:
> @ropg I don't have access to the hardware right now, is that the right fix?

Looks like it. Whoops. I'll be happy when this driver is actually tested on the hardware.

tore-espressif commented:
>More broadly: how bad is it if pin numbers are just an 'int' and that's it. Sure: you can set the wrong value. You can set lots of wrong values all over the project. If you have a wrong GPIO pin, it's either a typo or because you have a wrong idea of what GPIO pins something is connected to. In both cases the wrong value can just as easily be in the correct range, and then things still wouldn't work. How many people do we really help by enforcing a range here? Because there is significant cost: having to change code every time there's a new ESP32 out, just for the sake of a range in Kconfig

At seems to me that you are slowly getting the idea of getting rid of Kconfig :D 🍰 
Every ESP-IDF driver that uses GPIO pins checks whether the gpio index is valid for given peripheral (be it I2C, or GPIO) and for given chip (ESP32-xx). So we are just reinventing the wheel (not your fault, this is how it is and why I'd like to change it).

This is how I see it in this PR:
i2c_manager = Mutex + Convenience functions(read/write) + Kconfig

This is my idea of a future-proof solution:
i2c_manager = Mutex + Convenience functions + Runtime configuration

ropg commented:
I much appreciate thinking about this with you and others.

> This is how I see it in this PR:
> i2c_manager = Mutex + Convenience functions(read/write) + Kconfig
> 
> This is my idea of a future-proof solution:
> i2c_manager = Mutex + Convenience functions + Runtime configuration

How would that work assuming multiple components (example: LVGL and a 3D motion sensor thread) need I2C in terms of mutex vis-a-vis other ESP-IDF components? You'll need to share a single configuration for the I2C port(s) that is not specific to LVGL's runtime configuration (or LVGL period). If I were writing a cool 3D accelerometer thing, I would want to not invent the wheel and I'd look for something lightweight for thread-safety wrt various other users of I2C ports *like* LVGL. But I would not want to depend *on* LVGL, or even have to know it exists. What if that 3D accelerometer software, or the I2S audio amplifier/filter/ADC-thing, or ..., also has a runtime configuration that wants to set port parameters? Fight?

I agree (and had agreed already) that the move away from Kconfig makes total sense overall. I've though about it for a while, and feel that in the particular case of I2C the way the port works in ESP-IDF creates an unescapable need for a separate component and Kconfig is where you would configure the ports parameters (only). Bonus points if it's one that also has a mutex-compatible version that can be built into a bigger component so that not every user of something like LVGL needs to install it just because ESP-IDF lacks conditional depends. This sounds very much like I2C Manager to me.

Maybe I simply do not understand, but nothing I can envision under "Mutex + Convenience functions + Runtime configuration" would address the problems that I2C Manager now addresses. Maybe you can elaborate?

C47D commented:
I'm getting the following error message 

> E (20) i2c: i2c_param_config(644): i2c clock choice is invalid, please check flag and frequency
E (30) lvgl_i2c: Failed to initialise I2C port 0.
W (30) lvgl_i2c: If it was already open, we'll use it with whatever settings were used to open it. See I2C Manager README for details.
E (1050) i2c: i2c_set_pin(825): scl and sda gpio numbers are the same
W (1050) lvgl_i2c: Error writing addr 0x3c reg 0x00 port 0
assertion "0 == err" failed: file "../components/lvgl_esp32_drivers/lvgl_tft/ssd1306.c", line 133, function: ssd1306_init

Clock configuration is the default one, I'm setting pin 21 as SDA and 22 as SCL

C47D commented on i2c_manager/Kconfig:
```diff
@@ -0,0 +1,116 @@
+				range 0 21 if IDF_TARGET_ESP32C3
+			config I2C_MANAGER_0_FREQ_HZ
+				int "Frequency (Hz)"
+				default 1000000
```
According to [this](https://docs.espressif.com/projects/esp-idf/en/v4.1.1/api-reference/peripherals/i2c.html#_CPPv4N12i2c_config_t9clk_speedE) I2C clock speed can't be higher than 1_000_000, so i think setting this to 100_000 is safer, what do you think?

ropg replied:
Default now 400 kHz, not 1 MHz.

C47D commented on i2c_manager/Kconfig:
```diff
@@ -0,0 +1,116 @@
+			config I2C_MANAGER_0_FREQ_HZ
+				int "Frequency (Hz)"
+				default 1000000
+				range 100000 5000000
```
`5_000_000` is a bit to high according to the comment above.

C47D replied:
Yes, I think so.

C47D commented on i2c_manager/i2c_manager.c:
```diff
@@ -0,0 +1,321 @@
+
+		I2C_FN(_mutex)[port] = xSemaphoreCreateMutex();
+
+		i2c_config_t conf;
```
this variable needs to be zero initialized `i2c_config_t conf = {0}`.

ropg replied:
fixed

C47D reviewed (commented):
Here's some changes that were necessary to make the project compile.

C47D commented:
I tested the SSD1306 display without and with this PR, the former works, the latter doesn't, I will capture the I2C traffic tomorrow with a logic analyzer to try to spot the differences, the code looks a lot better with this PR tho, I feel it more maintainable, and certainly will help the developers when their projects requires more devices in the I2C bus 😸

C47D commented:
Here's the log of this PR:

```
Trace: lv_init started  (lv_obj.c #172 lv_init())
Info: lv_init ready     (lv_obj.c #231 lv_init())
I (10) lvgl_helpers: Display hor size: 128, ver size: 64
I (10) lvgl_helpers: Display buffer size: 1024
I (20) lvgl_helpers: init i2c display
I (30) lvgl_i2c: Starting I2C master at port 0.
I (30) lvgl_i2c: Initialised port 0 (SDA: 21, SCL: 22, speed: 400000 Hz.)
```

Here's the log of the working directory:
```
I (10) lvgl_helpers: Display hor size: 128, ver size: 64
I (10) lvgl_helpers: Display buffer size: 1024
I (20) lvgl_helpers: Initializing I2C master for display
I (30) lvgl_helpers: Initializing I2C master port 0...
I (30) lvgl_helpers: SDA pin: 21, SCL pin: 22, Speed: 400000 (Hz)
I (40) lvgl_helpers: Setting I2C master configuration...
I (50) lvgl_helpers: Installing I2C master driver...
Trace: Screen create started    (lv_obj.c #273 lv_obj_create())
Info: Object create ready       (lv_obj.c #461 lv_obj_create())
Trace: Screen create started    (lv_obj.c #273 lv_obj_create())
Info: Object create ready       (lv_obj.c #461 lv_obj_create())
Trace: Screen create started    (lv_obj.c #273 lv_obj_create())
Info: Object create ready       (lv_obj.c #461 lv_obj_create())
Trace: label create started     (lv_label.c #78 lv_label_create())
Trace: Object create started    (lv_obj.c #305 lv_obj_create())
Info: Object create ready       (lv_obj.c #461 lv_obj_create())
Info: label created     (lv_label.c #165 lv_label_create())
Trace: lv_task_handler started  (lv_task.c #67 lv_task_handler())
Trace: lv_refr_task: started    (lv_refr.c #177 _lv_disp_refr_task())
Trace: lv_refr_task: ready      (lv_refr.c #321 _lv_disp_refr_task())
Trace: lv_task_handler ready    (lv_task.c #180 lv_task_handler())
Trace: lv_task_handler started  (lv_task.c #67 lv_task_handler())
Trace: lv_task_handler ready    (lv_task.c #180 lv_task_handler())
```

C47D commented:
It's working now, the only change I needed was the one in the review. 😸

ropg commented:
@C47D Excellent. I'll rework the proposal later today or tomorrow.

@dastarling: I'll also rework your gt911 driver to use I2C Manager. Hope you'll be available to test as I do not have the hardware.

tore-espressif commented:
> How would that work assuming multiple components (example: LVGL and a 3D motion sensor thread) need I2C in terms of mutex vis-a-vis other ESP-IDF components?

Yes, that's a problem.
If we distribute I2C_Manager as part of LVGL, that it implies that **only** LVGL is supposed to use it. In other words: I2C_Manager _is part of_ LVGL. As opposed to LVGL _depends_ on I2C_Manager.

The former injects I2C_manager into your project, the latter requires I2C_manager to be in your project.

None of these options is ideal.

On the other hand, in current implementation, user can configure I2C_manager to only own I2C0 and then do whatever he wants with I2C1. This would reserve I2C0 for LVGL and leave I2C1 for the user.

> or ..., also has a runtime configuration that wants to set port parameters? Fight?

It shouldn't. It is a bad design if a component tries to configure an entity that is supposed to be shared.
Of course, it is often done this way, but then these components don't claim that they are portable and the user must guarantee that none else will access the shared entity.

> Maybe I simply do not understand, but nothing I can envision under "Mutex + Convenience functions + Runtime configuration" would address the problems that I2C Manager now addresses. Maybe you can elaborate?

* Mutex - guarantees thread safety
* Convenience functions - provide `i2c_read/write` functions so user doesn't have to copy esp-idf i2c calls all over the place
* Runtime configuration - configure I2C manager from application main function
I2C_Manager provides these three functionalities (currently with Kconfig instead of runtime configuration). Or is there anything else?

__Final note:__
ESP32 has two I2C peripherals that can be mapped to (almost) any GPIO pin. I find it reasonable that LVGL would reserve one of those for its purposes and let the other one for the user (sensors, or whatever). IMHO, there won't be too many users that would place touch driver, I2C display and 3D accelerometer on the same bus, when they can have two buses.
It would make everything a lot easier and more straightforward.
What do you think?

ropg commented:
@tore-espressif:

>On the other hand, in current implementation, user can configure I2C_manager to only own I2C0 and then do whatever he wants with I2C1. This would reserve I2C0 for LVGL and leave I2C1 for the user.

If the user, additional to LVGL with I2C Manager built-in installs I2C manager as a component they can make it configure and manage one or both ports, and use I2C Manager to talk to them. And with one statement, LVGL's copy and the standalone component sync up.

>The former injects I2C_manager into your project, the latter requires I2C_manager to be in your project. None of these options is ideal.

"Ideal" is a difficult term in a non-ideal world. But, IMHO, it is the optimum in a world where Espressif doesn't provide native thread-safety. It just gives the implementer both options, which are interoperable. Unless you see a third option, this is all it can do in terms of offering maximum flexibility.

> It shouldn't. It is a bad design if a component tries to configure an entity that is supposed to be shared.
Of course, it is often done this way, but then these components don't claim that they are portable and the user must guarantee that none else will access the shared entity.

Yes, and the unfortunate reality of the present is that people end up commenting out code from components that all initialize the same I2C port and end up without thread-safety.

> ESP32 has two I2C peripherals that can be mapped to (almost) any GPIO pin. I find it reasonable that LVGL would reserve one of those for its purposes and let the other one for the user (sensors, or whatever). IMHO, there won't be too many users that would place touch driver, I2C display and 3D accelerometer on the same bus, when they can have two buses.
It would make everything a lot easier and more straightforward.
What do you think?

I'm afraid it's not about whether you or me find something reasonable, it's about which devices are connected to what busses in actual hardware that's out there. In my M5Stack Core2, for instance, the 3D accelerometer and the touch sensor **are** connected to the same I2C bus. They both need to be read, very likely in different threads at different intervals, many times a second. Other devices I've seen use one port for all internal hardware and the other for externally wired / grove connector / extension / whatever-you-call-it purposes. When using I2C Manager none of this is a problem.

ropg commented:
I just added support for 10-bit addresses and 16-byte registers to I2C Manager.

@dastarling: I also updated the gt911 touch driver (which needed the 16-bit registers) to use I2C Manager. I cannot test as I do not have the device. But it compiles and I am sort of expecting things to work. Could you please test?

(Looking at the commit one can see how much easier the new driver I2C is to use.)

&nbsp;

**With these changes, I think I haven't forgotten anything and we're ready to merge...**

(I'll squash it to one commit before we actually merge. But let's test and all agree that this is it first.)

ropg commented on lvgl_touch/Kconfig:
```diff
@@ -98,7 +84,7 @@ menu "LVGL Touch controller"
             prompt "GPIO for MISO (Master In Slave Out)"
             range 0 39 if IDF_TARGET_ESP32
             range 0 43 if IDF_TARGET_ESP32S2
```
I made the range 0-46 for ESP32S2 everywhere.

ropg commented:
@tore-espressif @kisvegabor @embeddedt @C47D:

Naturally, when I actually sifted through the debate above I did come across some last loose ends: the commented-out code and the pin range that hadn't been reflected in code. But I really think we're all set for merge now. Last things:

* If've asked @dastarling, also via e-mail, to test the gt911 I2C touch sensor code. Haven't heard anything. That said I think the changes are very straightforward and I did make sure the new code compiles.

* If we all agree this is it, I'll squash the commits before the actual merge.

dastarling commented:
During testing, i only have the first I2C bus enabled and I am getting this error currently:

```
../components/lvgl_esp32_drivers/i2c_manager/i2c_manager.c: In function 'lvgl_i2c_init':
../components/lvgl_esp32_drivers/i2c_manager/i2c_manager.c:128:22: error: 'I2C_NUM_1' undeclared (first use in this function); did you mean 'I2C_NUM_0'?
   } else if (port == I2C_NUM_1) {
                      ^~~~~~~~~
                      I2C_NUM_0

```

dastarling commented:
While testing on the esp32c3 and GT911, I get an assert/abort()

```
E (1305) lvgl_i2c: I2C port 1 not configured.
ESP_ERROR_CHECK failed: esp_err_t 0xffffffff (ESP_FAIL) at 0x4038960d
file: "../components/lvgl_esp32_drivers/i2c_manager/i2c_manager.c" line 137
func: lvgl_i2c_init
expression: ESP_FAIL

abort() was called at PC 0x40389611 on core 0

```

ropg commented:
@dastarling Odd errors, I'll have a look later today.

dastarling commented:
I finally got it working, but had to make these 3 changes:

```
diff --git a/i2c_manager/i2c_manager.c b/i2c_manager/i2c_manager.c
index 0c05f76..5ed9e36 100644
--- a/i2c_manager/i2c_manager.c
+++ b/i2c_manager/i2c_manager.c
@@ -125,17 +125,17 @@ esp_err_t I2C_FN(_init)(i2c_port_t port) {
                                ESP_LOGE(TAG, "I2C port 0 not configured.");
                                ESP_ERROR_CHECK(ESP_FAIL);
                        #endif
+        #if defined (CONFIG_I2C_MANAGER_1_ENABLED)
                } else if (port == I2C_NUM_1) {
-                       #if defined (CONFIG_I2C_MANAGER_1_ENABLED)
                                conf.sda_io_num = CONFIG_I2C_MANAGER_1_SDA;
                                conf.scl_io_num = CONFIG_I2C_MANAGER_1_SCL;
                                conf.sda_pullup_en = I2C_MANAGER_1_PULLUPS ? GPIO_PULLUP_ENABLE : GPIO_PULLUP_DISABLE;
                                conf.scl_pullup_en = conf.sda_pullup_en;
                                conf.master.clk_speed = CONFIG_I2C_MANAGER_1_FREQ_HZ;
-                       #else
-                               ESP_LOGE(TAG, "I2C port 1 not configured.");
-                               ESP_ERROR_CHECK(ESP_FAIL);
-                       #endif
+        #else
+            ESP_LOGE(TAG, "I2C port 1 not configured.");
+//            ESP_ERROR_CHECK(ESP_FAIL);
+        #endif
                } else {
                        ESP_LOGE(TAG, "Invalid port.");
                        ESP_ERROR_CHECK(ESP_FAIL);
diff --git a/i2c_manager/i2c_manager.h b/i2c_manager/i2c_manager.h
index 6727cdc..eee8644 100644
--- a/i2c_manager/i2c_manager.h
+++ b/i2c_manager/i2c_manager.h
@@ -38,8 +38,8 @@ extern "C" {
 #define I2C_FN(s)         CONCAT(I2C_NAME_PREFIX, s)
 
 
-#define I2C_ADDR_10 ( 2^15 )
-#define I2C_REG_16  ( 2^31 )
+#define I2C_ADDR_10 ( 0x8000 )
+#define I2C_REG_16  ( 0x80000000 )
 
 esp_err_t I2C_FN(_init)(i2c_port_t port);
 esp_err_t I2C_FN(_read)(i2c_port_t port, uint16_t addr, uint32_t reg, uint8_t *buffer, uint16_t size);

```

This one evaluated to a value that caused it not to work I2C_ADDR_10 , for some reason when I printed it out it evaluated to 0x0D instead of the expected 0x08000.  Odd.

dastarling commented:
Alternately it could be:

#define I2C_ADDR_10 ( 1 << 15 )
#define I2C_REG_16  ( 1 << 31 )

dastarling commented:
Also you compare (Reg Address == 0) to determine if a register value is included.  But for some chips internal register 0 would be valid. I would suggest using -1 or 0xFFFFFFFF as the value for no register?

dastarling commented:
FYI ^ is exclusive OR so 2^15 is 13 not 0x8000.  I initially missed it as well.

ropg commented:
> FYI ^ is exclusive OR so 2^15 is 13 not 0x8000. I initially missed it as well.

Ah, wow, yes, DUH. Thank you.

ropg commented:
On the other issue: I have I2C port 1 disabled here also, which gives no trouble. I'm trying to figure out how your compiler could possibly complain it doesn't know about `I2C_NUM_1` ?  In `i2c_manager.h` I have `#include <driver/i2c.h>` (only to get these two defines) and that file defines them both:

```c
#define I2C_NUM_0              (0) /*!< I2C port 0 */
#define I2C_NUM_1              (1) /*!< I2C port 1 */
```

Could you check that your `[...]/esp-idf/components/driver/include/driver/i2c.h` gets included and that it has these two defines?

dastarling commented:
ESP32C3 only has 1 I2C, so there is no I2C_NUM_1 defined, so that is the issue.  perhaps you can use:

#define SOC_I2C_NUM            (1)

To determine how may I2C interfaces are available?

ropg commented:
> ESP32C3 only has 1 I2C, so there is no I2C_NUM_1 defined, so that is the issue. perhaps you can use:
> 
> #define SOC_I2C_NUM (1)
> 
> To determine how may I2C interfaces are available?


AH! Right!

Will do.

dastarling commented:
Retested latest version and it compile and works for GT911 on LVGL 7.11.0

C47D commented:
@ropg Do you want me to test it again? It seems like #84 caused a conflict

ropg commented:
@dastarling thanks

@C47D I will rebase in next few hours and make it one commit again.

ropg commented:
Alright, I rebased to current master to resolve the lvgl_touch/Kconfig conflict (I fixed indentation afterwards) and squashed everything to a single commit without conflicts.

HOWEVER....

When I compile my test code, which uses v7 release LVGL, I now get:

```
../components/lvgl_esp32_drivers/lvgl_tft/disp_spi.c: In function 'spi_ready':
../components/lvgl_esp32_drivers/lvgl_tft/disp_spi.c:313:33: error: incompatible type for argument 1 of 'lv_disp_flush_ready'
```

Which seems unrelated to anything my code does but a result my now having rebased to include #83. Am I doing something wrong or does that not work with v7 like it's intended to?

tore-espressif commented:
> In my M5Stack Core2, for instance,...

OK, this is a bulletproof argument, I agree.
Thank you very much for all your effort.

> does that not work with v7 like it's intended to...

LVGL v7 and v8 are unfortunately not compatible, so if you rebased on #83 then you have to support v8.

There were many commits since I last checked, so I'll go through it once again (tomorrow?), do some manual testing (still no CI 😿 ) and approve.

ropg commented:
@tore-espressif

> > [Core2]

> OK, this is a bulletproof argument, I agree.

It will only _really_ be the right decision if others start using I2C Manager also. I hope to stimulate uptake further so it can become the de-facto standard way for ESP32 software to do I2C. (I take suggestions for what other cool projects would benefit most...)

> LVGL v7 and v8 are unfortunately not compatible, so if you rebased on #83 then you have to support v8.

But it will take other changes to this repo before v8 'just works', or I am wrong?

Should we maybe have a branch that does v7 and I2C Manager as well as some of the newer drivers? After this merges that would simply be master minus a few commits. Even if it were allowed it to go stale alongside v7 over time, I think it'd be good to have for now. Shall I cherrypick and maintain for the time being? Is "LVGL_v7" a good branch name?

> There were many commits since I last checked, so I'll go through it once again (tomorrow?), do some manual testing (still no CI 😿 ) and approve.

Yay !

C47D commented:
@ropg We will need to add a check to solve that, something along the lines of:

```c
#if LVGL_VERSION_MAJOR < 8
        lv_disp_flush_ready(&disp->driver);
#else
        lv_disp_flush_ready(disp->driver);
#endif
```

tore-espressif commented:
> We will need to add a check to solve that, something along the lines of:

Alright, I was away for a while and when I saw #83 merged, I thought that you want to support v8 only.

> Github Actions for this project

Eventually yes, but there is a lot of work to be done before we get there🗡️ haha

ropg commented:
> @ropg We will need to add a check to solve that, something along the lines of:
> 
> ```c
> #if LVGL_VERSION_MAJOR < 8
>         lv_disp_flush_ready(&disp->driver);
> #else
>         lv_disp_flush_ready(disp->driver);
> #endif
> ```

Yes, I saw a check just like it for >= v7 right above that code but figured that must not be policy anymore. I just made that change and things work fine again with v7. I filed #86 with this change.

&nbsp;

**preprocessor directives alignment**

By the way, unrelated to the immediate issues at hand...

I know lots of people do it, but there is really no need to align preprocessor directives to the left. In my opinion code is almost always much more readable if the indentation just follows the decision flow, and compilers that needed them to be on the left are LOOOONG gone. (See [here](https://stackoverflow.com/questions/789073/indenting-defines) for details.)

After all: for reading code, it really doesn't matter if a decision is compiletime or runtime, your brain just wants to see the flow. Example: this is more readable than the alternative with the preprocessor stuff on the left, no? The difference gets even starker once things are nested.

```c
esp_err_t I2C_FN(_lock)(i2c_port_t port) {
    I2C_PORT_CHECK(port, ESP_FAIL);
    ESP_LOGV(TAG, "Mutex lock set for %d.", (int)port);

    TickType_t timeout;
    #if defined (I2C_ZERO)
        if (port == I2C_NUM_0) {
            timeout = (CONFIG_I2C_MANAGER_0_LOCK_TIMEOUT) / portTICK_RATE_MS;
        }
    #endif
    #if defined (I2C_ONE)
        if (port == I2C_NUM_1) {
            timeout = (CONFIG_I2C_MANAGER_1_LOCK_TIMEOUT) / portTICK_RATE_MS;
        }
    #endif

    if (xSemaphoreTake(I2C_FN(_mutex)[port], timeout) == pdTRUE) {
        return ESP_OK;
    } else {
        ESP_LOGE(TAG, "Removing stale mutex lock from port %d.", (int)port);
        I2C_FN(_force_unlock)(port);
        return (xSemaphoreTake(I2C_FN(_mutex)[port], timeout) == pdTRUE ? ESP_OK : ESP_FAIL);
    }
}
```

C47D commented:
> Alright, I was away for a while and when I saw #83 merged, I thought that you want to support v8 only.

>Yes, I saw a check just like it for >= v7 right above that code but figured that must not be policy anymore. I just made that change and things work fine again with v7. I filed #86 with this change.

My bad, I wrongly assumed the parameter were the same in v7 and v8.

> Eventually yes, but there is a lot of work to be done before we get there🗡️ haha

Indeed

> By the way, unrelated to the immediate issues at hand...
I know lots of people do it, but there is really no need to align preprocessor directives to the left. In my opinion code is almost always much more readable if the indentation just follows the decision flow, and compilers that needed them to be on the left are LOOOONG gone. (See here for details.)
After all: for reading code, it really doesn't matter if a decision is compiletime or runtime, your brain just wants to see the flow. Example: this is more readable than the alternative with the preprocessor stuff on the left, no? The difference gets even starker once things are nested.

I agree, there's a code formatting script hidden in the `lv_port_esp32` project (I think 😅), we should add this to it and let the tools automate code formatting, what do you think?

Thanks a lot for your work @ropg and @tore-espressif , I left this project unattended for a while and now it's showing that lack of attention to details.

ropg commented:
> I agree, there's a code formatting script hidden in the `lv_port_esp32` project (I think 😅), we should add this to it and let the tools automate code formatting, what do you think?

If you want to enforce formatting, I think it's best if the project publishes settings for one or more popular formatters so people can format their own code. You could then format again as part of testing/CI, but doing it this way makes it so most people can at least find stuff in their own code. Also this should not be set overly restrictive: E.g: I would still like to be able to occasionally write `if (condition) statement;` if it makes things more readable and I'm sure others have their sometimes-frowned-upon coding preferences too.

That said: it's surprisingly hard to find a set of rules that doesn't mess up some portion of code that is now perfectly readable. And if people contribute code that is maybe just lightly modified from somewhere else, enforcing formatting just makes it harder to maintain, so it's a bit of a double-edged sword. I think just having an agreed upon preferred format/settings and a policy to use this where possible but then being selective as to how/where to enforce it is most pragmatic for a project like this that seeks contributions for all sorts of hardware from a wider community. (Different rules could be applied to 'central' vs. 'driver' code maybe?)

tore-espressif commented on i2c_manager/i2c_manager.c:
```diff
@@ -0,0 +1,368 @@
+	TickType_t timeout = 0;
+	#if defined (I2C_ZERO)
+		if (port == I2C_NUM_0) {
+			timeout = (CONFIG_I2C_MANAGER_0_TIMEOUT) / portTICK_RATE_MS;
```
Could we use macros defined on lines 69 and 70?

ropg replied:
Yes, sorry, had missed this one...

tore-espressif commented on i2c_manager/i2c_manager.c:
```diff
@@ -0,0 +1,368 @@
+#include "i2c_manager.h"
+
+
+#if defined __has_include
+	#if __has_include ("esp_idf_version.h")
+		#include "esp_idf_version.h"
+		#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 3, 0)
+			#define HAS_CLK_FLAGS
+		#endif
+	#endif
+#endif
```
yes, the clk_flags in IDF 4.3 added some problems, but I think this solution is a little over-engineered.

This is what I was using and it compiles fine with or without `clk_flags` field in the config:

```
const i2c_config_t conf_master = {
        .mode = I2C_MODE_MASTER,
        .sda_pullup_en = GPIO_PULLUP_ENABLE,
        .scl_pullup_en = GPIO_PULLUP_ENABLE,
        .master.clk_speed = I2C_MASTER_FREQ_HZ,
        .sda_io_num = I2C_MASTER_SDA_IO,
        .scl_io_num = I2C_MASTER_SCL_IO
    };
```

The compiler will set all unset fields (clk_flags in IDF 4.3) to zero.

ropg replied:
If memory serves, it compiled without but then didn't work, giving me some obscure problem that I googled to be because of this.

tore-espressif replied:
alright, lets stick on the safe side

tore-espressif reviewed (approved):
Mostly details, otherwise LGTM!

Todos on my side:
- [ ] Check on HW

ropg commented on i2c_manager/Kconfig:
```diff
@@ -0,0 +1,116 @@
+		if I2C_MANAGER_0_ENABLED
+			config I2C_MANAGER_0_SDA
+				int "SDA (GPIO pin)"
+				range 0 39 if IDF_TARGET_ESP32
+				range 0 46 if IDF_TARGET_ESP32S2
+				range 0 21 if IDF_TARGET_ESP32C3
```
Shall I make it so for all I2C pins?

ropg replied:
Or even for all GPIO in the whole thing?

tore-espressif replied:
IMO, for all gpios. Let's assume that users know what they are doing. And in case they don't, esp-idf drivers have safety checks for this, so there will be no silent failures.

ropg replied:
I'll do it for I2C Manager here and then we'll do a separate PR later for the rest. We've got to finish this one at some point.

ropg commented on lvgl_tft/ssd1306.c:
```diff
@@ -240,17 +240,7 @@ static uint8_t send_pixels(lv_disp_drv_t *disp_drv, void *color_buffer, size_t b
-    /* Send queued commands */
-    err = i2c_master_cmd_begin(DISP_I2C_PORT, cmd, 10 / portTICK_PERIOD_MS);
-    i2c_cmd_link_delete(cmd);
+    err = lvgl_i2c_write(OLED_I2C_PORT, OLED_I2C_ADDRESS, OLED_CONTROL_BYTE_DATA_STREAM, color_buffer, buffer_len);
```
Does the LVGL display software care about this return value besides it being zero for OK or anything else for failure?

tore-espressif replied:
Nope, it just simplifies the code :D

ropg replied:
ok. Did same for `send_pixels` right below it.

ropg commented on i2c_manager/i2c_manager.c:
```diff
@@ -0,0 +1,368 @@
+		#define I2C_MANAGER_0_PULLUPS 	false
+	#endif
+
+	#define I2C_MANAGER_0_TIMEOUT 		CONFIG_I2C_MANAGER_0_TIMEOUT / portTICK_RATE_MS
+	#define I2C_MANAGER_0_LOCK_TIMEOUT	CONFIG_I2C_MANAGER_0_LOCK_TIMEOUT / portTICK_RATE_MS
```
You're right, will make it so.

ropg commented on i2c_manager/i2c_manager.c:
```diff
@@ -0,0 +1,368 @@
+		#define I2C_MANAGER_1_PULLUPS 	false
+	#endif
+
+	#define I2C_MANAGER_1_TIMEOUT 		CONFIG_I2C_MANAGER_1_TIMEOUT / portTICK_RATE_MS
+	#define I2C_MANAGER_1_LOCK_TIMEOUT	CONFIG_I2C_MANAGER_1_LOCK_TIMEOUT / portTICK_RATE_MS
```
You're right, will make it so.

ropg commented on i2c_manager/i2c_manager.h:
```diff
@@ -0,0 +1,76 @@
+#define I2C_FN(s) CONCAT(I2C_NAME_PREFIX, s)
+
+
+#define I2C_ADDR_10 ( 1 << 15 )
+#define I2C_REG_16  ( 1 << 31 )
+#define I2C_NO_REG  ( 1 << 30 )
```
Very much intended for users to supply. See README.

ropg commented:
Proposed way forward:

1. I will update everything I promise above, wait for response to the few open questions and then put that all in one additional commit.
2. I'll add another commit with a change I've made to the Kconfig, moving the outer menu to our project Kconfig  so we can use the same Kconfig file verbatim from the i2c_manager directory. It also has an improved and more correct LVGL I2C README file.
3. Last approval
4. Then I squash back to one commit
5. Go 


OK?

ropg commented:
I'll squash these back into one commit when you're done testing.

* I left in the 'overengineered' `HAS_CLK_FLAGS` because it did give me grief not to initialize that.

* I already noticed that the I2C README does not render pretty on GitHub, it does in my previewer. Will fix.

ropg commented:
As an aside, I added `i2cdev` support  I2C Manager. Doesn't affect what's in `lvgl_esp32_drivers`, but means that an additional ton of i2c drivers written for it can now run thread-safe with LVGL. (I2C Manager already emulated `I2C HAL`, the other abstraction layer.

tore-espressif commented:
> OK?

Pleasure to work with you :D 

> I'll add another commit with a change I've made to the Kconfig,

Good idea, it will imply that i2c_manager does not belong solely under LVGL, but should be share across the whole app

I'll do my best to check and do some basic tests by tomorrow.

tore-espressif commented on lvgl_tft/Kconfig:
```diff
@@ -1023,30 +996,33 @@ menu "LVGL TFT Display controller"
-            Configure the I2C SDA pin here.
+        config LV_I2C_DISPLAY_PORT_0
+            bool
+            prompt "I2C port 0"
```
Provide help about I2C ports
```suggestion
            prompt "I2C port 0"
            help
                I2C is shared peripheral managed by I2C_Manager. In order to configure I2C_manager (pinout, etc.) see menu
                Component config->LVGL ESP Drivers->I2C Port Settings.
```

tore-espressif commented on lvgl_tft/Kconfig:
```diff
@@ -1023,30 +996,33 @@ menu "LVGL TFT Display controller"
-            range 0 21 if IDF_TARGET_ESP32C3
+        config LV_I2C_DISPLAY_PORT_1
+            bool
+            prompt "I2C port 1"
```
Provide help about I2C ports
```suggestion
            prompt "I2C port 1"
            help
                I2C is shared peripheral managed by I2C_Manager. In order to configure I2C_manager (pinout, etc.) see menu
                Component config->LVGL ESP Drivers->I2C Port Settings.
```

tore-espressif commented on lvgl_touch/Kconfig:
```diff
@@ -548,4 +484,31 @@ menu "LVGL Touch controller"
+
+        config LV_I2C_TOUCH_PORT_0
+            bool
+            prompt "I2C port 0"
```
Provide help about I2C ports
```suggestion
            prompt "I2C port 0"
            help
                I2C is shared peripheral managed by I2C_Manager. In order to configure I2C_manager (pinout, etc.) see menu
                Component config->LVGL ESP Drivers->I2C Port Settings.
```

tore-espressif commented on lvgl_touch/Kconfig:
```diff
@@ -548,4 +484,31 @@ menu "LVGL Touch controller"
+
+        config LV_I2C_TOUCH_PORT_1
+            bool
+            prompt "I2C port 1"
```
Provide help about I2C ports
```suggestion
            prompt "I2C port 1"
            help
                I2C is shared peripheral managed by I2C_Manager. In order to configure I2C_manager (pinout, etc.) see menu
                Component config->LVGL ESP Drivers->I2C Port Settings.
```

tore-espressif reviewed (changes requested):
I hope that these are my last comments :D 
Apart from them, LGTM!

ropg commented on i2c_manager/README.md:
```diff
@@ -0,0 +1,90 @@
+&nbsp;
+
+
+## Information for LVGL driver developers
+
+I2C support in the LVGL ESP drivers is provided exclusively by the files in this directory. Code from all over the project that was talking to the I2C hardware directly has been replaced by code that communicates through the functions provided in `i2c_manager.h`. I2C is handled by the I2C Manager that was built into `lvlg_esp32_drivers`, but the code would be the same if it was routed through I2C ...
```
I'll try to think of more suitable text, but think we do need to say most of these things as it's not obvious otherwise.

tore-espressif replied:
yes, but in few months, when I2C_Manager is long merged, it won't be relevant anymore

ropg replied:
Right, that's true. Will fix.

ropg commented on Kconfig:
```diff
@@ -1,2 +1,14 @@
+
+    rsource "lvgl_touch/Kconfig"
+
+    menu "I2C Port Settings"
+        depends on LV_I2C && !HAVE_I2C_MANAGER
+
+        rsource "i2c_manager/Kconfig"
+    
+    endmenu
```
This menu disappears if I2C Manager the component is installed (`HAVE_I2C_MANAGER`).

In the odd case that it is not, but two components have it in 'OEM mode', they'll both show the menu showing the same settings (change in one immediately reflecting in other). But if you have to components that have I2C Manager you'd want to sync them up via the installed I2C Manager component, so this is not really an issue.

I guess it's fine to place it outside of the LVGL menu, I just figured it would be less confusing this way, but it may actually be better if they're always in the same place, even if there's an odd case where it's two menus.

ropg commented:
@tore-espressif 

Sorry about this late change...

Yesterday I noticed a logic error in how things work if components with I2C Manager depend on one another (something I had here). In that case `#include i2c_manager/i2c_manager.h` is not unique.

So I renamed the directory to `lvgl_i2c` (which fits our scheme better). I put that in a separate commit because it touches a lot of files and is easier to verify that way.

Also I noticed that i2c_manager would not have compiled under ESP-IDF 3.x, so I fixed that. (I think, haven't tested)

I'll do the other things you noticed today.

ropg commented:
Alright: I made your changes, looked at the menuconfig and tested with my hardware.

C47D commented:
Looks good to me!

@kisvegabor @embeddedt ANy thoughts?

embeddedt commented:
No comments.

There are a few extra diff lines that are just adding/deleting trailing whitespace, but those can be filtered easily when looking at the diff down the road, so I don't personally think they matter.

C47D commented:
Ok, I will merge this awesome job, thanks a lot @ropg 😸

ropg commented:
Woot!

This has to have been one of the longer threads under a pull request, but everything was needed and things were so much improved. It has been a great pleasure working with you, and we have a better lvgl_esp32_drivers as well as a better I2C Manager to show for it.

ropg commented:
( Just a pity I didn't squash the commits before the final merge, which would have made the commit history more readable since it now looks out of order because things happened in between my partial commits. But this works... )
